src/Microsoft.Framework.DependencyInjection.Abstractions/ServiceCollectionExtensions.cs

```diff
@@ -232,6 +231,39 @@ public static bool TryAddSingleton<TService>([NotNull] this IServiceCollection c
             return true;
         }
 
+        /// <summary>
+        /// Adds the specified <see cref="ServiceDescriptor"/>s if an existing descriptor with the same
+        /// <see cref="ServiceDescriptor.ServiceType"/> and an implementation that does not already exist
+        /// in <paramref name="services."/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="descriptor">The <see cref="ServiceDescriptor"/>s.</param>
+        /// </param>
+        /// <returns>
+        /// <c>true</c> if any of the <paramref name="descriptor"/>s are added to the collection, otherwise <c>false</c>.
+        /// </returns>
+        /// <remarks>
+        /// Use <see cref="TryAddEnumerable"/> when registing a service implementation of a service type that
+        /// supports multiple registrations of the same service type. Using
+        /// <see cref="Add(IServiceCollection, ServiceDescriptor)"/> is not idempotent and can add duplicate
+        /// <see cref="ServiceDescriptor"/> instances if called twice. Using
+        /// <see cref="TryAddEnumerable(IServiceCollection, ServiceDescriptor)"/> will prevent registration
+        /// of multiple implementation types.
+        /// </remarks>>
+        public static bool TryAddEnumerable(
```

Comment:
I'm really tempted to make this method and the `TryAdd` equivalent void. If no one is really using a return value, we shouldn't return something that isn't intuitive.